Sources/PackageSyntax/PackageEditor.swift

```diff
@@ -0,0 +1,427 @@
+                    }
+                } else {
+                    context.diagnosticsEngine.emit(.missingProductOrTarget(name: dependency))
+                    throw Diagnostics.fatalError
+                }
+            }
+        case .binary(name: let name, urlOrPath: let urlOrPath, checksum: let checksum):
+            guard loadedManifest.toolsVersion >= .v5_3 else {
+                context.diagnosticsEngine.emit(.unsupportedToolsVersionForBinaryTargets)
+                throw Diagnostics.fatalError
+            }
+            try editor.addBinaryTarget(targetName: name, urlOrPath: urlOrPath, checksum: checksum)
+        }
+
+        try context.verifyEditedManifest(contents: editor.editedManifest)
+        try fs.writeFileContents(manifestPath, bytes: ByteString(encodingAsUTF8: editor.editedManifest))
+
+        // Write template files.
+        try writeTemplateFilesForTarget(newTarget)
+
+        if case .library(name: let name, includeTestTarget: true, dependencyNames: _) = newTarget {
+            try self.addTarget(.test(name: "\(name)Tests", dependencyNames: [name]),
+                               productPackageNameMapping: productPackageNameMapping)
+        }
+    }
+
+    private func diagnoseUnsupportedToolsVersions(manifest: Manifest) throws {
+        guard manifest.toolsVersion >= .v5_2 else {
+            context.diagnosticsEngine.emit(.unsupportedToolsVersionForEditing)
+            throw Diagnostics.fatalError
+        }
+    }
+
+    private func writeTemplateFilesForTarget(_ newTarget: NewTarget) throws {
+        switch newTarget {
+        case .library:
+            let targetPath = context.manifestPath.parentDirectory.appending(components: "Sources", newTarget.name)
+            if !localFileSystem.exists(targetPath) {
+                let file = targetPath.appending(component: "\(newTarget.name).swift")
+                try fs.createDirectory(targetPath, recursive: true)
+                try fs.writeFileContents(file, bytes: "")
+            }
+        case .executable:
+            let targetPath = context.manifestPath.parentDirectory.appending(components: "Sources", newTarget.name)
+            if !localFileSystem.exists(targetPath) {
+                let file = targetPath.appending(component: "main.swift")
+                try fs.createDirectory(targetPath, recursive: true)
+                try fs.writeFileContents(file, bytes: "")
+            }
+        case .test:
+            let testTargetPath = context.manifestPath.parentDirectory.appending(components: "Tests", newTarget.name)
+            if !fs.exists(testTargetPath) {
+                let file = testTargetPath.appending(components: newTarget.name + ".swift")
+                try fs.createDirectory(testTargetPath, recursive: true)
+                try fs.writeFileContents(file) {
+                    $0 <<< """
+                    import XCTest
```

Comment:
Over time we should find a way to unify this with the package init (in fact, it might be reasonable to rewrite that as package editing operations), but this seems alright for this PR.

Reply:
yeah, agreed!